### PR TITLE
Cost history enhancements

### PR DESCRIPTION
**Pull request by nkuitse**
I've broken ajax_*.php down into separate files, one for each case in the three big switch statements.

### CODE REVIEW

**Comment by benheet** [comment]
We've passed our two week window on this one. Are we ready to merge it in?

**Comment by jeffnm** [comment]
I have not tested this, but if it's passing other people's tests, then merge away. There seem to be some conflicts in the pull request though, so someone should probably look into that.

**Comment by benheet** [comment]
Paul H., can you review and address the conflicts that Jeffrey mentioned?

**Comment by jeffnm** [comment]
We are testing this after all. I've installed it on our test sever and our e-resources librarian will be taking a look at it soon. I'll report back any issues we encounter.

**Comment by PaulPoulain** [comment]
After applying this patch & updating the database:
 * I loose existing acquisition types. Maybe it was expected, but it's a pain, I had to re-enter them. On a live install, that would be a big problem
 * I can now enter a resource. With the parameter set to Y I see the ENH cost screen:
capture du 2015-02-20 10 30 42

There are 2 problems though:
 * the dates are in US format : february 1st is displayed 2/1/2015 (which is unreadable for most of the world ;-) )
 * when you want to enter more than one line, the new lines have the date-picker not working : only the 1st datepicker work well (the one at the bottom)

And one question/comment : the % line is not very clear. I entered the following datas:
capture du 2015-02-20 10 35 02

I expected the % to be calculated only on the "same" line of "last year", ie the same fund/same type line (70€ became 100€ the year after. Instead, the +71 mean that the total cost of 2015 is 71% more than the total cost of 2014)
I think that we should split this % increase in another table, for a better clarity. Like:
2014: 100       n/a
2015: 220     +71.4%

**Comment by PaulPoulain** [comment]
Note for other testers, here is how I tested this patch.
starting from a fresh 1.2 release (retrieved from coral github):
 cd $CORAL/resources
# adding another remote source:
git remote add flo https://github.com/fenway-libraries-online/coral-resources.git

#fetching it
git remote update

#checkout the branch flo
git checkout flo
(git log show me that there are patches from paul@flo.org, fine)

# preparing the DB update and running it
perl -i -pe "s/_DATABASE_NAME_/coral_resources_prod/g" install/protected/update_1.2_flo.sql
mysql coral_resources_prod < install/protected/update_1.2_flo.sql
(no error output, nice)

# updating configuration
vi admin/configuration.ini
adding  enhancedCostHistory=Y line
# restarting apache:
/etc/init.d/apache2 restart

**Comment by PaulPoulain** [comment]
Another problem : the DB update. It's too complex, without reason.
The lines containing things like
ALTER TABLE `_DATABASE_NAME_`.`Resource`
must be changed to be
ALTER TABLE `Resource`

the DB name will be provided when the file is run on mysql:
mysql -u login -p password CORAL_DB <update.sql, so the __DATABASE_NAME__ is useless

**Comment by nkuitse** [comment]
How do I see the conflicts?

**Comment by nkuitse** [comment]
Dates in US format area silly oversight on my part.  :-(  Display date format should IMO be a config file setting.  I'll look into that.

**Comment by nkuitse** [comment]
That's bizarre that you lost acquisition types.  I'll fix that.  I'll also remove the spurious `DATABASE_NAME`. bits.

**Comment by nkuitse** [comment]
Finally, the percent increase could be implemented in different ways; we chose to make it reflect the change in total expenditures from year to year, regardless of fund.  However, when we presented the enhancement to our own internal customers -- the ERM librarians who worked on the spec -- they decided that they didn't want a percentage increase display after all, because they found it distracting and unhelpful.  So perhaps the solution is to drop it from the enhancement.

**Comment by PaulPoulain** [comment]
Le 20/02/2015 22:56, Paul Hoffman a écrit :
>
> Dates in US format area silly oversight on my part. :-( Display date 
> format should IMO be a config file setting. I'll look into that.
>
>
I think that there's PHP way of knowing the system format date that 
could help: date_default_timezone_set. Isn't there an option in the date 
controler about that ?

PS : when coming into date problems, the best solution is usually to 
deal everything in ISO internally (YYYY-MM-DD), and switch to local 
format only when displaying.

-- 
Paul Poulain, Associé-gérant / co-owner
BibLibre, expert du logiciel libre pour les bibliothèques
BibLibre, Open Source software for libraries expert

**Comment by PaulPoulain** [comment]
Le 20/02/2015 23:01, Paul Hoffman a écrit :
>
> Finally, the percent increase could be implemented in different ways; 
> we chose to make it reflect the change in total expenditures from year 
> to year, regardless of fund. However, when we presented the 
> enhancement to our own internal customers -- the ERM librarians who 
> worked on the spec -- they decided that they didn't want a percentage 
> increase display after all, because they found it distracting and 
> unhelpful. So perhaps the solution is to drop it from the enhancement.
>
I agree with the idea of dropping it for now, and maybe have another 
board with statistic numbers to see how things evolve over time.

-- 
Paul Poulain, Associé-gérant / co-owner
BibLibre, expert du logiciel libre pour les bibliothèques
BibLibre, Open Source software for libraries expert

**Comment by nkuitse** [comment]
Could someone please tell me how to view the conflicts in this pull request?  I haven't been able to figure this out.

**Comment by jeffnm** [comment]
Here's some info about resolving conflicts https://help.github.com/articles/resolving-a-merge-conflict-from-the-command-line/

**Comment by nkuitse** [comment]
That info is for committers; it doesn't help me.

**Comment by jeffnm** [comment]
The commit conflicts that are preventing an automatic merge are what I was
referring to previously. If there is some other conflict, I'm not sure how
to help with that.

**Comment by nkuitse** [comment]
Right -- and thanks for pointing me to the info you found -- but I don't know how to see which commit conflicts are preventing an automatic merge.  I'm guessing that I need to clone the master repo anew, then attempt to merge my commits, but I can't quite get my head around how to do that.

**Comment by jeffnm** [comment]
I see what you mean. I guess changes were made on the master branch between
the time you checked it out originally and your pull request.

This page shows you the commits that the Master Resource module has
received since you forked it.
https://github.com/ndlibersa/resources/compare/fenway-libraries-online:flo...ndlibersa:master

This page shows the commits and changes you've made on your fork that
aren't in the master yet.
https://github.com/fenway-libraries-online/coral-resources/compare/ndlibersa:master...master

Anywhere the same lines are changed or a file has been deleted will need to
be manually approved. This can be done by the person who does the final
merge, or you can pull the changes from the ndlibersa:master branch into
your branch, fix any problems, and then make another pull request.

I hope that helps.

**Comment by nkuitse** [comment]
I have fixed the conflicts and successfully rebased onto the latest HEAD (commit 357fe9fe89).  Or at least I believe that is so; GitHub still says that the pull request contains merge conflicts.  (Do I need to submit a new pull request???)  I have also fixed the problem with the DB upgrade SQL.  Could someone please check to see if this looks OK before I proceed with fixing the date problem and deleting the percent change code?  

Finally, I would feel better having this enhancement committed onto a separate branch and then merged into master; is there some reason not to create a branch for significant changes such as this?

**Comment by remocrevo** [comment]
I'd suggest that any major development should be merged by hand, not by using GitHub's auto-merge feature.  You could comment on the GitHub issue and add a link to your latest branch.  Then one or two people can test your branch and comment on the issue that they approve (or any concerns they have).  When a few people are happy with it, someone can pull your branch to their own machine, merge it with master, and push it to the main GitHub repository.  At that point it is considered "released" to the community, so we would need to change the version number, provide information about installing and using the changes, announce to the community, etc.

Does that sound like a reasonable workflow to everyone?

**Comment by nkuitse** [comment]
That sounds good to me.  Here's a link:

https://github.com/fenway-libraries-online/coral-resources

The latest branch is master.

**Comment by nkuitse** [comment]
I'm looking at the SQL code for upgrading to 1.1 and 1.2 (install/protected/update_1.1.sql and install/protected/update_1.2.sql) and they both have `_DATABASE_NAME_` prepended to every table name.  The upgrade script (install/update.php) substitutes the actual database name for each occurrence of this placeholder. So IMO this should stay in the DB upgrade code for 1.3.

**Comment by nkuitse** [comment]
The format_date function defined in directory.php has had the format "m/d/Y" hard-coded since commit a110e2c four years ago (or earlier).  The PHP docs say (http://php.net/manual/en/function.date.php) to use setlocale() and strftime() for locale-sensitive dates -- which seems entirely correct to me, but I wouldn't want to try to make that change in this upcoming release.

**Comment by nkuitse** [comment]
Finally, I don't see how the cost history enhancement upgrade could delete acquisition types -- the DB upgrade SQL code doesn't touch the AcquisitionType table.  Paul, are you sure that they've been deleted?  Or are they merely omitted in the display?

**Comment by remocrevo** [comment]
It sounds like @nkuitse has some questions we need to answer before this code can be merged, especially about the problems @PaulPoulain experienced during his testing. @PaulPoulain can you confirm that the Acquisition Type is still getting lost?  And are the other problems fixed (e.g. date picker)?

**Comment by nkuitse** [comment]
Sorry, I totally forgot about the broken date picker -- I'll look into that now.  My three most recent comments address the other problems @PaulPoulain mentioned: the _DATABASE_NAME_ prefix and date formatting are just as they were before, and I don't see any way that our changes could have deleted acquisition types (though certainly they may have messed up the *display* of acquisition types).

**Comment by nkuitse** [comment]
FYI, I've successfully upgraded two of our member libraries' production CORAL instances from version 1.2 to 1.3 (this version, with our enhancements).  I pulled the master branch from our GitHub repo, added the enhancedCostHistory=Y line to admin/configuration.ini, visited the install page (/resources/install) in my browser, clicked on Run Update 1.3, and that was that -- no problems.  (They have a couple of very minor customizations -- e-mail sender and alert text -- which I isolated in a local branch; then I merged master into the local branch.  I went slowly and carefully and it took about 20 minutes each, very smooth, no tweaking or fiddling required.)

**Comment by benheet** [comment]
Sounds like this is getting close. That's awesome. Has the missing acquisition type problem that Paul Poulain pointed out been dealt with too? Since this hasn't been merged yet it would be nice if the date picker would be fixed too before this is final.

**Comment by PaulPoulain** [comment]
I tried to test again this patch, and the update_1.3.sql file still contain the _DATABASE_NAME_ pragma, that is useless (see previous comment), please remove it.

(will test again after. I don't understand what I made wrong, and I don't understand why acq types have disappeared)

**Comment by nkuitse** [comment]
_DATABASE_NAME_ is not useless; the upgrade script uses it:

    //Process the sql file by statements
    foreach ($statements as $statement) {
       if (strlen(trim($statement))>3){
            //replace the DATABASE_NAME parameter with what was actually input
            $statement = str_replace("_DATABASE_NAME_", $installer->getDatabaseName(), $statement);
    
            $result = mysql_query($statement);
            if (!$result){
                $installer->addErrorMessage(mysql_error() . "<br /><br />For statement: " . $statement);
                break;
            }
        }
    }

**Comment by remocrevo** [comment]
From my quick investigation, it looks like the _DATABASE_NAME_ prefix could be removed from the SQL file without causing harm to the installation/upgrade process. But I would want someone to test that. In general, I think there are better ways for accessing a specific database, such as the mysql_select_db() function that CORAL already uses. So I am happy for us to remove the prefix and the code that uses it. But that is not a necessary change right now, and it would slow down the process of releasing this code. I recommend that @PaulPoulain create a github "issue" for that change so we can address it separately, and if it is causing a lot of trouble for testing the db scripts, we can address it quickly.

**Comment by nkuitse** [comment]
I've been going around in circles trying to fix the date picker bug that Paul Poulain spotted in the "Edit cost history" box. I figured out that I have to re-add date pickers when the user clicks on the "Add" button, so that the added row will get a date picker for each of its two date fields:

    $(".addPayment").live('click', function () { 
        // 150+ lines omitted
        $('.date-pick').datePicker({startDate:'01/01/1996'});
        return false;
    }

And indeed when I make that change works, in a way -- it adds a date picker to every date field, which means that the date fields that already existed on the page now have two date pickers.  :-(

So I tried this, which should (I think) remove the date pickers; now the code looks like this:

    $(".addPayment").live('click', function () { 
        // 150+ lines omitted
        $('.dp-choose-date').remove();
        $('.date-pick').datePicker({startDate:'01/01/1996'});
        return false;
    }

And indeed it does that, sort of -- the date pickers are removed, and the only date pickers are in the newly added row.

Any ideas how to fix this?  Should we just go ahead and merge the pull request with this bug present?

**Comment by nkuitse** [comment]
Never mind, I figured it out.  The fix is a hack, though -- it requires duplicating some code from jquery.datepicker, because the version we're using doesn't have the functionality we need.  Here's the final addition just FTR; I'll commit the change in a minute:

    // Remove all date pickers
    $('.dp-choose-date').remove();
    // HACK: clear the internal ID
    $('.date-pick').each(function() {
            this._dpId = 0;
    });
    // HACK: clear the cache (may not be necessary)
    var els = $.event._dpCache || [];
    for (var i in els) {
            $(els[i].ele)._dpDestroy();
    }
    $.event._dpCache = [];
    // Re-add all date pickers
    $('.date-pick').datePicker({startDate:'01/01/1996'});
    return false;

**Comment by nkuitse** [comment]
Remington, are you OK with merging this now?

File: install/manual_upgrade_1_3.txt [hunk 1]
```diff
@@ -0,0 +1,20 @@
+Manual Installation Coral Upgrade from 1.2 to 1.3
+
+Update the resources module codebase
+
+	-When performing upgrades your configuration file (/admin/configuration.ini) 
+	should always be backed up first so that when you check out the new code you are sure to not overwrite it. 
+	-Check out the new code through GitHub https://github.com/ndlibersa/resources/
+	-If needed manually copy and overwrite all the files into the exiting resources directory.
+	-Do not replace the existing directory.  This will cause you to lose any settings, documents, etc. That you may have.  Copying the new files over the existing files and replacing them will ensure you get the changes needed but not removing additional files.
+	-Ensure that your your configuration file (/admin/configuration.ini) is still correct.
```
Comment (remocrevo) [inline install/manual_upgrade_1_3.txt]: Duplicate word "your your", one should be removed.

**Comment by remocrevo** [comment]
Hm, the date picker fix does look pretty hacky. I'm impressed that you figured it out, but I think it's best if we can find a better way. So my colleague and I installed this branch, tested it, and found a slightly simpler fix. Tell me what you think. You should be able to replace your whole change with the following four lines:

// Remove datepickers from clone
originalTR.find('.dp-choose-date').remove();
// Re-add datepickers to clone
originalTR.find('.date-pick').datePicker({startDate:'01/01/1996'});

Looking into it a little more, it seems we are on a very old version of jQuery, and an even older version of the date picker plugin. Ultimately, we need to upgrade CORAL to a modern version of jQuery and make sure we keep up with new versions in the future. But for now, I will create a separate issue for that.

**Comment by nkuitse** [comment]
I've committed your revised fix.  I agree on postponing a jQuery upgrade for now, FWIW.

**Comment by remocrevo** [comment]
Also, regarding @PaulPoulain 's problem of losing acquisition types, maybe that was a misunderstanding? In our testing, we didn't lose those values, but we noticed they were moved to a separate edit form. And the new cost history form contains an empty "Cost Details" dropdown which might have been the cause of confusion. @PaulPoulain does that explain the problem you had? It might help to include some default values with this branch. I'm not really sure what the intended use of that dropdown is. @nkuitse can you comment on that and/or supply some default entries for the CostDetails table?

**Comment by nkuitse** [comment]
The spec we worked up for the enhancement says the new CostDetails table "provides a controlled entry list for use at the institution's discretion" and my reading of things is that it's a way to give people a flexible (but controlled) way to categorize subscriptions or payments that is orthogonal to the payment type and fund, and that avoids shoehorning stuff into free-text notes.  The only examples given in the spec are "FTE", "FTE 25", and "FTE 26".  I've asked Kelly for clarification in case she has any additional insights -- she was deeply involved in the writing of the spec, which I was not.

**Comment by remocrevo** [comment]
I still want to make sure @PaulPoulain wasn't experiencing a different bug. If he can confirm that his data wasn't actually lost, then we can prepare to merge. I'll try to catch him on IRC on Monday.

**Comment by remocrevo** [comment]
@nkuitse I think we still need some user documentation for this branch. I see the upgrade instructions, but nothing about how to use the new features.

**Comment by benheet** [comment]
@nkuitse you might be able to use the original spec to update existing user guide

**Comment by nkuitse** [comment]
Where is the existing user guide?  Our docs for 1.3 are here:

https://docs.google.com/document/d/1tFLgc55hXSoU7aQ-pHjm09JAQaO40cZcIQ64LuGFb7Q/edit?usp=sharing

I've generated a PDF of this but don't see where to put it.

**Comment by PaulPoulain** [comment]
OK, I tried to test again, and, good news, the acquisition type empty does not come from the patch. It came from my sample database, that had it empty. Sorry for the useless warning.

About the datepicker, I'm not sure I understand: applying only this patch, I still have the dates in US format. But I should ignore the problem, because it will be fixed by a jquery upgrade ? If yes, I'm fine with that, you can merge this patch imo.

**Comment by nkuitse** [comment]
I'm glad to hear that the acquisition types weren't deleted.  As for the US-centric display of dates -- it's not right, but it seems that's the way dates have always been displayed in CORAL Resources so at least we're not making matters worse.

**Comment by remocrevo** [comment]
I have added a GitHub issue for improving the display of dates (issue #38). Otherwise, I believe this branch is ready to be merged into master. Thanks @nkuitse for the updated user guide. The PDF could be added to the CORAL website once we get login accounts figured out. @PaulPoulain are you able to work on that? Anything you need from us?

**Comment by PaulPoulain** [comment]
am I able to work on that => you mean date display ?
Yes, definetly, that's something I'll work on with @veggiematts and our intern @Cesarfr

**Comment by PaulPoulain** [comment]
(and maybe the jquery update will help us fixing this annoying problem)

**Comment by benheet** [comment]
Remington, do you want to go ahead and do the merge? We can post the updated guide to the discussion listserv until it has a permanent home on the website.

**Comment by benheet** [comment]
@nkuitse can you email me, bjheet@gmail.com, your PDF copy of the updated resources module user guide? We'll replace the existing one on the coral website with it. 

@remocrevo would you mind handling the merge? I think we are ready.

@nkuitse you may want to start crafting an announcement and get ready to send it to coral-erm@listserv.nd.edu. I think its past time to get this one finished and out there.

**Comment by nkuitse** [comment]
I've just sent Ben the PDF.  We'll work on an announcement.  Thanks again!

**Pull request merged**